sparse/src/KokkosSparse_CooMatrix.hpp

```diff
@@ -0,0 +1,157 @@
+  using memory_space = typename device_type::memory_space;
+  //! Type of the Kokkos::MemoryTraits
+  using memory_traits = MemoryTraits;
+  //! Type of all integral class members
+  using size_type = SizeType;
+
+  static_assert(std::is_integral_v<OrdinalType>,
+                "OrdinalType must be an integral.");
+
+  //! The type of the row index view in the matrix
+  using row_view =
+      Kokkos::View<row_type*, Kokkos::LayoutRight, device_type, memory_traits>;
+  //! The type of the column index view in the matrix
+  using column_view = Kokkos::View<column_type*, Kokkos::LayoutRight,
+                                   device_type, memory_traits>;
+  //! The type of the scalar values view in the matrix
+  using scalar_view = Kokkos::View<scalar_type*, Kokkos::LayoutRight,
+                                   device_type, memory_traits>;
+  //! The type of a constant CooMatrix
+  using const_type = CooMatrix<const_scalar_type, const_ordinal_type,
+                               device_type, memory_traits, size_type>;
+
+ private:
+  size_type m_num_rows, m_num_cols;
+
+ public:
+  //! The row indexes of the matrix
+  row_view row;
+  //! The column indexes of the matrix
+  column_view col;
+  //! The scalar values of the matrix
+  scalar_view data;
+
+  /// \brief Default constructor; constructs an empty sparse matrix.
+  KOKKOS_INLINE_FUNCTION
+  CooMatrix() : m_num_rows(0), m_num_cols(0) {}
+
+  // clang-format off
+  /// \brief Constructor that accepts a column indicies view, row indices view, and
+  ///        values view.
+  ///
+  /// The matrix will store and use the column indices, rows indices, and values
+  /// directly (by view, not by deep copy).
+  ///
+  /// \param nrows   [in] The number of rows.
+  /// \param ncols   [in] The number of columns.
+  /// \param row_in  [in] The row indexes.
+  /// \param col_in  [in] The column indexes.
+  /// \param data_in [in] The values.
+  // clang-format on
+  CooMatrix(size_type nrows, size_type ncols, row_view row_in,
+            column_view col_in, scalar_view data_in)
+      : m_num_rows(nrows),
+        m_num_cols(ncols),
+        row(row_in),
+        col(col_in),
+        data(data_in) {
+    if (data.extent(0) != row.extent(0) || row.extent(0) != col.extent(0)) {
+      std::ostringstream os;
+      os << "data.extent(0): " << data.extent(0) << " != "
+         << "row.extent(0): " << row.extent(0) << " != "
+         << "col.extent(0): " << col.extent(0) << ".";
+      KokkosKernels::Impl::throw_runtime_exception(os.str());
+    }
+  }
+
+  //! The number of columns in the sparse matrix.
+  KOKKOS_INLINE_FUNCTION size_type numCols() const { return m_num_cols; }
+
+  //! The number of rows in the sparse matrix.
+  KOKKOS_INLINE_FUNCTION size_type numRows() const { return m_num_rows; }
+
+  //! The number of stored entries in the sparse matrix, including zeros.
+  KOKKOS_INLINE_FUNCTION size_type nnz() const { return data.extent(0); }
+};
+
```

Comment:
I expected to see methods to get row. col, and data. Otherwise I don't know how to use the COO data structure.